### PR TITLE
fix: runInBackground always run after setImmediate

### PR DESCRIPTION
**Pull request by dead-horse**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

##### Description of change

runInBackground will run after setImmedate

### CODE REVIEW

File: app/extend/context.js [hunk 1]
```diff
@@ -222,8 +222,10 @@ const proto = module.exports = {
     const start = Date.now();
     /* istanbul ignore next */
     const taskName = scope._name || scope.name || eggUtils.getCalleeFromStack(true);
-    // use app.toAsyncFunction to support both generator function and async function
-    return ctx.app.toAsyncFunction(scope)(ctx)
+    // use setImmediate to ensure all sync logic will run async
+    return new Promise(resolve => setImmediate(resolve))
```
Comment (popomore) [inline app/extend/context.js:226]: 这个改动有些单测可能会挂

File: test/fixtures/apps/ctx-background/app/controller/sync.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+module.exports = async ctx => {
+  const start = Date.now();
+  ctx.runInBackground(async () => {
+    const start = Date.now();
+    while(Date.now() - start < 100) {
+    }
```
Comment (fengmk2) [inline test/fixtures/apps/ctx-background/app/controller/sync.js:8]: 加一个 runInBackground 结束的断言

File: app/extend/context.js (hunk 1 above)
Comment (dead-horse) [reply app/extend/context.js:226]: egg 这边的单测应该都过了，一般也没人依赖这个执行顺序吧？

**Comment by dead-horse** [comment]
这个我先合掉了，测试单独开个 PR 来修复

**Pull request merged**